Framework/Geometry/src/Instrument/StructuredDetector.cpp

```diff
@@ -0,0 +1,701 @@
+        "StructuredDetector::initialize(): x.size() should be = y.size()");
+  if (x.size() != ((m_xpixels + 1) * (m_ypixels + 1)))
+    throw std::invalid_argument("StructuredDetector::initialize(): x.size() "
+                                "should be = (xpixels+1)*(ypixels+1)");
+
+  // Store vertices
+  m_xvalues = x;
+  m_yvalues = y;
+
+  std::string name = this->getName();
+  auto minDetId = idstart;
+  auto maxDetId = idstart;
+  auto detSize = x.size();
+
+  CompAssembly *detAssembly = new CompAssembly(name);
+
+  for (auto ix = 0; ix < m_xpixels; ix++) {
+    // Create an ICompAssembly for each x-column
+    std::ostringstream oss_col;
+
+    oss_col << name << "(x=" << ix << ")";
+    CompAssembly *xColumn = new CompAssembly(oss_col.str(), this);
+
+    for (auto iy = 0; iy < m_ypixels; iy++) {
+      std::ostringstream oss;
+      oss << name << "(" << ix << "," << iy << ")";
+
+      // Calculate its id and set it.
+      auto id = this->getDetectorIDAtXY(ix, iy);
+
+      // Store min and max det IDs
+      // minimum structured detector id
+      if (id < minDetId) {
+        minDetId = id;
+      }
+      // maximum structured detector id
+      if (id > maxDetId) {
+        maxDetId = id;
+      }
+
+      // Create and store detector pixel
+      auto detector = addDetector(xColumn, oss.str(), ix, iy, id);
+      xColumn->add(detector);
+    }
+  }
+
+  m_minDetId = minDetId;
+  m_maxDetId = maxDetId;
+}
+
+/** Creates new hexahedral detector pixel at row x column y using the
+*   detector vertex values.
+* @param name :: The Pixel name identifier
+* @param x :: The pixel row
+* @param y :: The pixel column
+* @param id :: The pixel ID
+* @ returns newly created detector.
+*/
+Detector *StructuredDetector::addDetector(CompAssembly *parent,
+                                          const std::string &name, int x, int y,
+                                          int id) {
+  auto w = m_xpixels + 1;
+  std::ostringstream shapestr;
+  Mantid::Geometry::ShapeFactory shapeCreator;
+
+  // Store hexahedral vertices for detector shape
+  auto xlb = m_xvalues[(y * w) + x];
+  auto xlf = m_xvalues[(y * w) + x + w];
+  auto xrf = m_xvalues[(y * w) + x + w + 1];
+  auto xrb = m_xvalues[(y * w) + x + 1];
+  auto ylb = m_yvalues[(y * w) + x];
+  auto ylf = m_yvalues[(y * w) + x + w];
+  auto yrf = m_yvalues[(y * w) + x + w + 1];
+  auto yrb = m_yvalues[(y * w) + x + 1];
+
+  // calculate midpoint of trapeziod
+  auto a = abs(xrf - xlf);
+  auto b = abs(xrb - xlb);
+  auto h = abs(ylb - ylf);
+  auto cx = ((a + b) / 4);
+  auto cy = h / 2;
+
+  // store detector position before translating to origin
+  auto xpos = xlb + cx;
+  auto ypos = ylb + cy;
+
+  // Translate detector shape to origin
+  xlf -= xpos;
+  xrf -= xpos;
+  xrb -= xpos;
+  xlb -= xpos;
+  ylf -= ypos;
+  yrf -= ypos;
+  yrb -= ypos;
+  ylb -= ypos;
+
+  // Create XML shape used to describe detector pixel
+  shapestr << "<type name=\"userShape\" >";
```

Comment:
Typically for things like this I tend to use the [BuilderPattern](https://en.wikipedia.org/wiki/Builder_pattern). I think this would need to be generated by hand, but you could create a `HexahedronShapeBuilder`, which might be used like:

```cpp
HexahedronShapeBuilder hexBuilder;
hexBuilder.addLeftBackBottomPoint(x, y, z);
hexBuilder.addRightBackTopPoint(x, y, z);
// .....
boost::shared_ptr<Mantid::Geometry::Object> shape = hexBuilder.create();

```
* `HexahedronShapeBuilder` would internally call the  `Mantid::Geometry::ShapeFactory` on create()
* `HexahedronShapeBuilder` should check that all points have been added before creating anything
* `HexahedronShapeBuilder` should check that the points are correctly defined. i.e. the top should not have a `y` less than the `y` of the bottom. etc.